pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/generator/RealtimeToOfflineSegmentsTaskGenerator.java

```diff
@@ -0,0 +1,238 @@
+
+      // Fetch RealtimeToOfflineSegmentsTaskMetadata ZNode for reading watermark
+      RealtimeToOfflineSegmentsTaskMetadata realtimeToOfflineSegmentsTaskMetadata =
+          _clusterInfoProvider.getMinionRealtimeToOfflineSegmentsTaskMetadata(realtimeTableName);
+
+      if (realtimeToOfflineSegmentsTaskMetadata == null) {
+        // No ZNode exists. Cold-start.
+        long watermarkMillis;
+
+        String startTimeStr = taskConfigs.get(RealtimeToOfflineSegmentsTask.START_TIME_MILLIS_KEY);
+        if (startTimeStr != null) {
+          // Use startTime config if provided in taskConfigs
+          watermarkMillis = Long.parseLong(startTimeStr);
+        } else {
+          // Find the smallest time from all segments
+          RealtimeSegmentZKMetadata minSegmentZkMetadata = null;
+          for (LLCRealtimeSegmentZKMetadata realtimeSegmentZKMetadata : completedSegmentsMetadataList) {
+            if (minSegmentZkMetadata == null || realtimeSegmentZKMetadata.getStartTime() < minSegmentZkMetadata
+                .getStartTime()) {
+              minSegmentZkMetadata = realtimeSegmentZKMetadata;
+            }
+          }
+          Preconditions.checkState(minSegmentZkMetadata != null);
+
+          // Convert the segment minTime to millis
+          long minSegmentStartTimeMillis =
+              minSegmentZkMetadata.getTimeUnit().toMillis(minSegmentZkMetadata.getStartTime());
+
+          // Round off according to the bucket. This ensures we align the offline segments to proper time boundaries
+          // For example, if start time millis is 20200813T12:34:59, we want to create the first segment for window [20200813, 20200814)
+          watermarkMillis = (minSegmentStartTimeMillis / bucketMillis) * bucketMillis;
+        }
+
+        // Create RealtimeToOfflineSegmentsTaskMetadata ZNode using watermark calculated above
+        realtimeToOfflineSegmentsTaskMetadata =
+            new RealtimeToOfflineSegmentsTaskMetadata(realtimeTableName, watermarkMillis);
+        _clusterInfoProvider.setRealtimeToOfflineSegmentsTaskMetadata(realtimeToOfflineSegmentsTaskMetadata);
+      }
+
+      // WindowStart = watermark. WindowEnd = windowStart + bucket.
+      long windowStartMillis = realtimeToOfflineSegmentsTaskMetadata.getWatermarkMillis();
+      long windowEndMillis = windowStartMillis + bucketMillis;
+
+      // Check that execution window is older than bufferTime
+      if (windowEndMillis > System.currentTimeMillis() - bufferMillis) {
+        LOGGER.info(
+            "Window with start: {} and end: {} is not older than buffer time: {} configured as {} ago. Skipping scheduling task: {}",
+            windowStartMillis, windowEndMillis, bufferMillis, bufferTimeStr, taskType);
+      }
+
+      // Find all COMPLETED segments with data overlapping execution window: windowStart (inclusive) to windowEnd (exclusive)
+      List<String> segmentNames = new ArrayList<>();
```

Comment:
Instead of two lists, maybe you can have another list of LLCRealtimeSegmentZkMetadata?

Reply:
I did not follow. these lists are for setting comma separated segments and downloadURL into the task configs